File: Day 2/Part 2.php
<?php
// Luke Sheppard
// Advent of Code 2019 - Day 2 - Part 2 - https://adventofcode.com/2019/day/2#part2

$input = file_get_contents(__DIR__ . '/input.csv');

$target = 19690720;

for ($noun = 0; $noun < 100; $noun++) {
    for ($verb = 0; $verb < 100; $verb++) {
        $opcodes = explode(',', $input);
        $opcodes[1] = $noun;
        $opcodes[2] = $verb;

        for ($i = 0; $i < count($opcodes); $i += 4) {
            $opcode = $opcodes[$i];
            $params = array_slice($opcodes, $i + 1, 3);

            switch($opcode) {
                case 99: // Halt
                    break 2;

                case 1: // Add
                    $opcodes[$params[2]] = $opcodes[$params[0]] + $opcodes[$params[1]];
                    break;

                case 2: // Multiply
                    $opcodes[$params[2]] = $opcodes[$params[0]] * $opcodes[$params[1]];
                    break;
            }
        }

        if ($opcodes[0] == $target) {
            break 2;
        }
    }
}


echo "Answer: " . (100 * $noun + $verb) .  "\n";

File: Day 2/Part 2.rb
input = File.open(__dir__ + '/input.csv').read

target = 19690720

answer = catch :finish do
    100.times do |noun|
        100.times do |verb|
            opcodes = input.split(",").map(&:to_i)
            opcodes[1] = noun
            opcodes[2] = verb

            opcodes.each_slice(4).with_index do |slice, index|
                opcode = slice[0]

                break if opcode == 99

                case opcode
                when 1
                    opcodes[slice[3]] = opcodes[slice[1]] + opcodes[slice[2]]
                when 2
                    opcodes[slice[3]] = opcodes[slice[1]] * opcodes[slice[2]]
                end

            end

            throw :finish, (100 * noun + verb) if opcodes[0] == target
        end
    end
end

puts "Answer: #{answer}"
